### ISSUE TITLE
CI Test gets stuck in an infinite loop

### ISSUE THREAD

**Issue by ssinad** [ivadomed/ivadomed#778]
## Issue description
testing/functional_tests/test_automate_training.py does not terminate on macOS and Windows CI environments.

### Current behavior

The test gets stuck.

### Expected behavior
 
The test should terminate within minutes.

### Steps to reproduce

https://github.com/ssinad/ivadomed/actions/runs/814749769

### System description

macOS 10.15, windows-latest

### Installed packages

<details open>
<summary> Output of <code>pip freeze</code> </summary>

```

```

</details>

**Comment by dyt811** [comment]
Yeah... that is odd. 

Even within IvadoMed main source repo, we don't have multi OS CI yet. A PR in progress (#757) is trying to add that and I had a lot of fun with macOS CI specifically not complying etc but I yet to run into infinite loop issues. I have not examined your specific multi CI setup but given your recent merge from the main fork and assuming no major changes/alterations, this is a bit concerning if the source of problem stem from recent PR merges to the main repositories. 

Not sure if you can try my PR branch and see if that get you a bit further on CI?

**Comment by konantian** [comment]
@ssinad I have met your issue before, and there was a matplotlib window that popped up and I had to close that window manually to continue the testing process. Can you confirm if this is the same as your situation?

**Comment by dyt811** [comment]
I noticed  that in my CI GitHub script PR, I have some commands that might have gotten around this and it is not in yours. Maybe try adding that and see if it makes any differences, if this is indeed related to interactivity? 

https://github.com/ivadomed/ivadomed/pull/757/files#diff-c2e9f7c1bfaf65da79bbd29ed3ce389b2acc8b4099a257914a7292a66e6b9bc9R15

![image](https://user-images.githubusercontent.com/6720068/117254141-09b06200-ae16-11eb-849d-1c473b0ad13f.png)

**Comment by ssinad** [comment]
> I noticed that in my CI GitHub script PR, I have some commands that might have gotten around this (as my PR is tracking master pretty closely and did not hang for windows and mac at least during my dev testings) and it is not in yours. Maybe try adding that and see if it makes any differences, if the problem is indeed related to interactivity?
> 
> https://github.com/ivadomed/ivadomed/pull/757/files#diff-c2e9f7c1bfaf65da79bbd29ed3ce389b2acc8b4099a257914a7292a66e6b9bc9R15
> 
> ![image](https://user-images.githubusercontent.com/6720068/117254141-09b06200-ae16-11eb-849d-1c473b0ad13f.png)

You mean the DEBIAN_FRONTEND environment variable?  
I added it, but it didn't help.  
https://github.com/ssinad/ivadomed/actions/runs/816946026

**Comment by dyt811** [comment]
TLDR: pytest will not terminate because HDF5 environmental variable HDF5_USE_FILE_LOCKING='FALSE' not set. 

Further replications and detailed examination of CI was able to narrow this to be a regression of the earlier reported issue in https://github.com/ivadomed/ivadomed/issues/633 and https://github.com/ivadomed/ivadomed/issues/768 but finally reproduced in the CI context. Yes, we need to kill HDF5 **soon** before this trip people more time in the upcoming days. 

This cause issues in 10.15 macOS and latest macOS. We are not sure why Windows CI was cancelled but so far we have not been able to reproduce that issue so we will keep an eye out post the multi OS CI build PR #757 is fully incorporated before close this issue.

**Comment by ssinad** [comment]
So, I guess it's safe to close this issue.

**Comment by dyt811** [comment]
Apparently not out of the wood yet. Latest PR once fixed the missing fixture "pytest-console-scripts" actually causes hanging on both windows and Mac. This now fully replicate your original source problem and is VERY bizarre and almost single-handedly point the stuck issue to test_automated_training. 

PR before fixing that missing fixture error during collection but past the REST of the code 100%: See [here](https://github.com/ivadomed/ivadomed/runs/2551167941?check_suite_focus=true#step:7:21). 

PR after that fix adding it back, [stuck for hours and die](https://github.com/ivadomed/ivadomed/pull/757/checks?check_run_id=2551184035).

**Comment by dyt811** [comment]
Hey, @mpompolas, your earlier PR#572 adding the Violin plot MAY be causing our CI process to hang due to out of focus during the pytest of automate_training process. I was able to reproduce locally on Mac and PC and unless we close that window, the pytest hangs. Do you have a way to use environmental variable to suppress the actual showing of the graph while not affecting the functionalities? Any help greatly appreciated.

**Comment by mpompolas** [comment]
https://github.com/ivadomed/ivadomed/blob/master/ivadomed/scripts/automate_training.py#L473

Add new flag `plot_comparison=False`

```
def automate_training(file_config, file_config_hyper, fixed_split, all_combin, path_data=None,
                      n_iterations=1, run_test=False, plot_comparison=False, all_logs=False, thr_increment=None,
                      multi_params=False, output_dir=None):
```


https://github.com/ivadomed/ivadomed/blob/master/ivadomed/scripts/automate_training.py#L778

Substitute the flag that controls the plotting (I had it connected to `run_test` flag since that already did a comparison of the models) with `plot_comparison`

```
if plot_comparison:
        output_folders = [config_list[i]["path_output"] for i in range(len(config_list))]
        violin_plots.visualize_and_compare_models(ofolders=output_folders)
```


Let me know if that solves the issue, and if not I can dig into it more

**Comment by dyt811** [comment]
@mpompolas Awesome. Right on target. Issue resolved. CI no longer blocked locally or on [GitHubAction](https://github.com/ivadomed/ivadomed/pull/757/checks?check_run_id=2572188169).

@ssinad since you have been playing with CI much before, I would really use your feedback on [this rather important CI PR](https://github.com/ivadomed/ivadomed/pull/757/). Take a look and let me know what you think. Along with multi OS CI, the cascade separation of the requirements.txt to support both conda environmental and pip while also allowing dev/GPU requirements to share as much in common as possible. 

Once merged, PR#757 should close this issue automatically.
